Source/WebCore/Modules/cookie-store/CookieStore.cpp

```diff
@@ -56,25 +63,291 @@ namespace WebCore {
 
 WTF_MAKE_ISO_ALLOCATED_IMPL(CookieStore);
 
-Ref<CookieStore> CookieStore::create(Document* document)
+class CookieStore::MainThreadBridge : public ThreadSafeRefCounted<MainThreadBridge, WTF::DestructionThread::Main> {
+public:
+    static Ref<MainThreadBridge> create(CookieStore& cookieStore)
+    {
+        return adoptRef(*new MainThreadBridge(cookieStore));
+    }
+
+    void get(CookieStoreGetOptions&&, int promiseIdentifier);
+    void getAll(CookieStoreGetOptions&&, int promiseIdentifier, SecurityOrigin*, bool globalObjectIsWindowObject);
+    void set(CookieInit&& options, int promiseIdentifier);
+
+private:
+    MainThreadBridge(CookieStore&);
+
+    void ensureOnMainThread(Function<void(ScriptExecutionContext& context)>&&);
+    void ensureOnContextThread(ScriptExecutionContextIdentifier, Function<void()>&&);
+
+    WeakPtr<CookieStore, WeakPtrImplWithEventTargetData> m_cookieStore;
+};
+
+CookieStore::MainThreadBridge::MainThreadBridge(CookieStore& cookieStore)
+    : m_cookieStore(cookieStore)
 {
-    auto cookieStore = adoptRef(*new CookieStore(document));
-    cookieStore->suspendIfNeeded();
-    return cookieStore;
 }
 
-CookieStore::CookieStore(Document* document)
-    : ActiveDOMObject(document)
+void CookieStore::MainThreadBridge::ensureOnMainThread(Function<void(ScriptExecutionContext&)>&& task)
 {
-    if (!document)
+    ASSERT(m_cookieStore);
+    if (!m_cookieStore)
         return;
 
-    auto* page = document->page();
-    if (!page)
+    auto* context = m_cookieStore->scriptExecutionContext();
+    if (!context)
+        return;
+    ASSERT(context->isContextThread());
+
+    Ref protectedThis { *this };
+    if (auto* document = dynamicDowncast<Document>(*context)) {
+        task(*context);
+        return;
+    }
+
+    downcast<WorkerGlobalScope>(*context).thread().workerLoaderProxy().postTaskToLoader(WTFMove(task));
+}
+
+void CookieStore::MainThreadBridge::ensureOnContextThread(ScriptExecutionContextIdentifier contextIdentifier, Function<void()>&& task)
+{
+    ScriptExecutionContext::ensureOnContextThread(contextIdentifier, WTFMove(task));
+}
+
+void CookieStore::MainThreadBridge::get(CookieStoreGetOptions&& options, int promiseIdentifier)
```

Comment:
It would be nicer if this function took a `CompletionHandler<void(std::optional<Vector<Cookie>>&&)>` instead of a promiseIdentifier. Same comment applies to other MainThreadBride function. I think promise settling should stay in CookieStore, not in the bridge.